File: Lecture9Exercise/nullridgepgs_vs_pca.R
library(glmnet)
library(ggplot2)

simulate_and_correlate <- function(N, M = 500, structure_strength = 2.0) {
  # 1. Create a "True" ancestry gradient for the individuals
  ancestry <- seq(-1, 1, length.out = N)
  
  # 2. Assign SNP 'loadings' (how much each SNP reflects ancestry)
  # Most SNPs are noise, but some carry the ancestry signal
  loadings <- rnorm(M, 0, 1)
  
  # 3. Generate Genotype Matrix X: (Ancestry * Loadings) + Random Noise
  # This creates a matrix where PC1 is a clear "Ancestry" signal
  X <- outer(ancestry, loadings) * structure_strength + matrix(rnorm(N * M), N, M)
  X <- scale(X) # Standardize
  
  # 4. Extract the top Principal Component (PC1)
  pc_res <- prcomp(X, rank. = 1)
  pc1 <- pc_res$x[, 1]
  
  # 5. Generate a NULL phenotype (pure random noise, NO genetic effect)
  y <- rnorm(N)
  
  # 6. Fit Ridge Regression (alpha = 0)
  # Use cross-validation to find the "best" lambda
  cv_fit <- cv.glmnet(X, y, alpha = 0)
  
  # 7. Calculate the PGS for these individuals
  pgs <- as.vector(predict(cv_fit, X, s = "lambda.min"))
  
  # 8. Return the absolute correlation between the Null PGS and PC1
  return(abs(cor(pgs, pc1)))
}

# --- Execution ---
n_sizes <- c(100, 250, 500, 1000, 2500)
results <- data.frame()

for (n in n_sizes) {
  message(paste("Running simulation for N =", n))
  # Run 5 iterations per sample size to get an average
  corrs <- replicate(5, simulate_and_correlate(n))
  results <- rbind(results, data.frame(N = n, Mean_Corr = mean(corrs), SD = sd(corrs)))
}

# --- Plotting ---
ggplot(results, aes(x = N, y = Mean_Corr)) +
  geom(line() + 
  geom_point(size = 3) +
  geom_errorbar(aes(ymin = Mean_Corr - SD, ymax = Mean_Corr + SD), width = 50) +
  labs(title = "Correlation of Null PGS with Ancestry (PC1)",
       subtitle = "As Sample Size (N) increases, the 'Spurious' Ancestry signal decreases",
       x = "Sample Size (N)",
       y = "Abs(Correlation: PGS vs PC1)") +
  theme_minimal()

File: Lecture9Exercise/nullridgepgs_vs_pca.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn.linear_model import RidgeCV
from sklearn.preprocessing import StandardScaler

def simulate_and_correlate(N, M=500, structure_strength=2.0):
    # 1. Generate population structure
    # Latent ancestry gradient
    ancestry = np.linspace(-1, 1, N)
    # SNPs have different 'loadings' on this ancestry
    loadings = np.random.normal(0, 1, M)
    # Genotype matrix: Structure + Noise
    # We add structure to about 20% of SNPs to make a clear PC1
    X = np.outer(ancestry, loadings) * structure_strength + np.random.normal(0, 1, (N, M))
    
    # 2. Standardize genotypes
    scaler = StandardScaler()
    X_std = scaler.fit_transform(X)
    
    # 3. Get PC1
    pca = PCA(n_components=1)
    pc1 = pca.fit_transform(X_std).flatten()
    
    # 4. Generate NULL phenotype (random noise, no relation to X or ancestry)
    y = np.random.normal(0, 1, N)
    
    # 5. Fit Ridge Regression (PGS)
    # RidgeCV automatically finds the best alpha (lambda)
    model = RidgeCV(alphas=np.logspace(-2, 5, 10))
    model.fit(X_std, y)
    pgs = model.predict(X_std)
    
    # 6. Calculate absolute correlation
    correlation = np.abs(np.corrcoef(pgs, pc1)[0, 1])
    return correlation

# Range of sample sizes
n_sizes = [100, 250, 500, 1000, 2500, 5000]
results = []
errors = []

for n in n_sizes:
    iterations = 10
    corrs = [simulate_and_correlate(n) for _ in range(iterations)]
    results.append(np.mean(corrs))
    errors.append(np.std(corrs))

# Create the plot
plt.errorbar(n_sizes, results, yerr=errors, fmt='-o', capsize=5, color='darkblue', label='Abs(Corr(PGS, PC1))')
plt.axhline(y=0, color='black', linestyle='--', alpha=0.3)
plt.xlabel('Sample Size (N)')
plt.ylabel('Correlation with PC1')
plt.title('Effect of Sample Size on Null PGS Ancestry Correlation')
plt.grid(True, linestyle=':', alpha=0.6)
plt.legend()
plt.savefig('null_pgs_sample_size.png')

# Output data for confirmation
results_df = pd.DataFrame({'N': n_sizes, 'Mean_Corr': results, 'Std_Dev': errors})
print(results_df)
